compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/ConstructorProcessing.kt

```diff
@@ -161,6 +162,38 @@ private fun processConstructors(
             //TODO: why don't we use declared member scope at this point?
             scope?.processDeclaredConstructors {
                 if (includeInnerConstructors || !it.fir.isInner) {
+                    // If the matched symbol is a type alias, and the expanded type is a nested class, e.g.,
+                    //
+                    //   class Outer {
+                    //     inner class Inner
+                    //   }
+                    //   typealias OI = Outer.Inner
+                    //   fun foo() { Outer().OI() }
+                    //
+                    // the chances are that `processor` belongs to [ScopeTowerLevel] (to resolve type aliases at top-level), which treats
+                    // the explicit receiver (`Outer()`) as an extension receiver, whereas the constructor of the nested class may regard
+                    // the same explicit receiver as a dispatch receiver (hence inconsistent receiver).
+                    // Here, we add a copy of the nested class constructor, along with the outer type as an extension receiver, so that it
+                    // can be seen as if resolving:
+                    //
+                    //   fun Outer.OI(): OI = ...
+                    //
+                    if (it.callableId.classId?.isNestedClass == true && matchedSymbol is FirTypeAliasSymbol) {
+                        val outerClassId = it.callableId.classId?.outerClassId!!
+                        val outerClass = session.firProvider.getFirClassifierByFqName(outerClassId) as? FirClass<*>
+                        val outerType = outerClass?.getPrimaryConstructorIfAny()?.returnTypeRef
+                        if (outerType != null) {
+                            val extCopy = buildConstructorCopy(it.fir) {
+                                origin = FirDeclarationOrigin.FakeOverride
```

Comment:
Probably, it's worth having a different kind of origin here

Reply:
How about Synthetic? :)